src/xr/xr-depth-sensing.js

```diff
@@ -0,0 +1,281 @@
+ * //
+ * //     // get luminance alpha components from depth texture
+ * //     vec2 packedDepth = texture2D(texture_depthSensingMap, texCoord).ra;
+ * //
+ * //     // unpack into single value in millimeters
+ * //     float depth = dot(packedDepth, vec2(255.0, 256.0 * 255.0)); // mm
+ * //
+ * //     // normalize: 0m to 8m distance
+ * //     depth = min(depth / 8000.0, 1.0); // 0..1 = 0m..8m
+ * //
+ * //     // paint scene from black to white based on distance
+ * //     gl_FragColor = vec4(depth, depth, depth, 1.0);
+ * // }
+ */
+function XrDepthSensing(manager) {
+    EventHandler.call(this);
+
+    this._manager = manager;
+    this._depthInfo = null;
+    this._available = false;
+
+    this._matrixDirty = false;
+    this._matrix = new Mat4();
+    this._emptyBuffer = new Uint8Array(32);
+    this._depthBuffer = null;
+
+    this._texture = new Texture(this._manager.app.graphicsDevice, {
+        format: PIXELFORMAT_L8_A8,
+        mipmaps: false,
+        addressU: ADDRESS_CLAMP_TO_EDGE,
+        addressV: ADDRESS_CLAMP_TO_EDGE
+    });
+
+    this._manager.on('end', this._onSessionEnd, this);
+}
+XrDepthSensing.prototype = Object.create(EventHandler.prototype);
+XrDepthSensing.prototype.constructor = XrDepthSensing;
+
+/**
+ * @event
+ * @name pc.XrDepthSensing#available
+ * @description Fired when depth sensing data becomes available.
+ */
+
+/**
+ * @event
+ * @name pc.XrDepthSensing#unavailable
+ * @description Fired when depth sensing data becomes unavailable.
+ */
+
+/**
+ * @event
+ * @name pc.XrDepthSensing#resize
+ * @description Fired when depth sensing texture been resized. So {@link pc.XrDepthSensing#uvMatrix} needs to be updated for relevant shaders.
+ * @param {number} width - The new width of the depth texture in pixels.
+ * @param {number} height - The new height of the depth texture in pixels.
+ * @example
+ * depthSensing.on('resize', function () {
+ *     material.setParameter('matrix_depth_uv', depthSensing.uvMatrix);
+ * });
+ */
+
+XrDepthSensing.prototype._onSessionEnd = function () {
+    this._depthInfo = null;
+
+    if (this._available) {
+        this._available = false;
+        this.fire('unavailable');
+    }
+
+    this._depthBuffer = null;
+    this._texture._width = 4;
+    this._texture._height = 4;
+    this._texture._levels[0] = this._emptyBuffer;
+    this._texture.upload();
+};
+
+XrDepthSensing.prototype._updateTexture = function () {
+    if (this._depthInfo) {
+        var resized = false;
+
+        // changed resolution
+        if (this._depthInfo.width !== this._texture.width || this._depthInfo.height !== this._texture.height) {
+            this._texture._width = this._depthInfo.width;
+            this._texture._height = this._depthInfo.height;
+            this._matrixDirty = true;
+            resized = true;
+        }
+
+        var dataBuffer = this._depthInfo.data;
+        this._depthBuffer = new Uint8Array(dataBuffer.buffer, dataBuffer.byteOffset, dataBuffer.byteLength);
```

Comment:
Presumably this is done every frame. Can we avoid this typed array creation here. I know this is just a buffer view and not a full new array, but still, can we avoid it? The `data` property is a `Uint16Array` - can it not be set on the texture directly?

Reply:
This was raised question in the specs: https://github.com/immersive-web/depth-sensing/issues/10
And there is a PR in the works to revisit this: https://github.com/immersive-web/depth-sensing/pull/11
There is an allocation done by UA, to provide software path to the depth data. Then we have a way now to create view (not allocating new typed array), which then pushed to GPU. What is in the works currently, is an approach to optimize it to either CPU path or GPU path, which will avoid unnecessary copying. But for now we have to do it that way.

Reply:
OK, let's just remember to revisit this in the future. 😄